utils/nctl/sh/scenarios/sync_upgrade_test.sh

```diff
@@ -0,0 +1,220 @@
+#!/usr/bin/env bash
+
+source "$NCTL"/sh/utils/main.sh
+source "$NCTL"/sh/views/utils.sh
+source "$NCTL"/sh/assets/upgrade.sh
+source "$NCTL"/sh/node/svc_"$NCTL_DAEMON_TYPE".sh
+
+# Exit if any of the commands fail.
+set -e
+
+#######################################
+# Runs an integration tests that tries to sync a new node
+# with an upgraded network.
+#
+# Arguments:
+#   `node=XXX` ID of a new node. Default=6
+#   `timeout=XXX` timeout (in seconds) when syncing. Default=300 seconds.
+#   `version=X_Y_Z` new protocol version to upgrade to. Default=2_0_0.
+#   `era=X` at which the upgrade should take place. Default=3.
+#######################################
+function main() {
+    log "------------------------------------------------------------"
+    log "Syncing node begins"
+    log "------------------------------------------------------------"
+
+    do_await_genesis_era_to_complete
+
+    # 1. Send batch of Wasm deploys
+    do_send_wasm_deploys
+    # 2. Send batch of native transfers
+    do_send_transfers
+    # 3. Wait until they're all included in the chain.
+    do_await_deploy_inclusion
+    # 4. Upgrade the network
+    do_upgrade_network
+    # 5. Wait for the network to upgrade.
+    do_await_network_upgrade
+    # 6. Take a note of the last finalized block hash
+    do_read_lfb_hash
+    # 7. Send batch of Wasm deploys
+    do_send_wasm_deploys
+    # 8. Send batch of native transfers
+    do_send_transfers
+    # 9. Wait until they're all included in the chain.
+    do_await_deploy_inclusion
+    # 10. Start the node in sync mode using hash from 4)
+    do_start_new_node "$NEW_NODE_ID"
+    # 11. Wait until it's synchronized
+    #     and verify that its last finalized block matches other nodes'.
+    do_await_full_synchronization "$NEW_NODE_ID"
+
+    log "------------------------------------------------------------"
+    log "Syncing node complete"
+    log "------------------------------------------------------------"
+}
+
+function log_step() {
+    local COMMENT=${1}
+    log "------------------------------------------------------------"
+    log "STEP $STEP: $COMMENT"
+    log "------------------------------------------------------------"
+    STEP=$((STEP + 1))
+}
+
+function do_await_genesis_era_to_complete() {
+    log_step "awaiting genesis era to complete"
+    while [ "$(get_chain_era)" != "1" ]; do
+        sleep 1.0
+    done
+}
+
+function do_upgrade_network() {
+    log_step "scheduling the network upgrade to version ${PROTOCOL_VERSION} at era ${ACTIVATE_ERA}"
+    for NODE_ID in $(seq 1 "$(get_count_of_nodes)"); do
+        _upgrade_node "$PROTOCOL_VERSION" "$ACTIVATE_ERA" "$NODE_ID"
+    done
+}
+
+function do_await_network_upgrade() {
```

Comment:
maybe we should add a timeout or an error check here so we don't potentially wait forever in case of a bug

Reply:
Oh yeah, definitely a good idea :) especially that I hit this problem in my tests :sweat_smile:

Reply:
Done in [`36f81a6` (#972)](https://github.com/CasperLabs/casper-node/pull/972/commits/36f81a604ef7657252e0912ae46bc115a5681649)